src/v/datalake/translation/tests/partition_translator_tests.cc

```diff
@@ -0,0 +1,357 @@
+        if (_test_ctx.error_on_translation()) {
+            throw std::runtime_error("translation error simulation");
+        }
+        if (batches.empty()) {
+            co_return;
+        }
+        auto min = model::offset_cast(batches.begin()->base_offset());
+        auto max = model::offset_cast(batches.back().last_offset());
+
+        // keep track of translation boundaries for result propagation
+        _min_offset_translated = std::min(
+          min, _min_offset_translated.value_or(kafka::offset::max()));
+        _max_offset_translated = std::max(
+          max, _max_offset_translated.value_or(kafka::offset::min()));
+    }
+
+    ss::future<> flush() final {
+        if (_test_ctx.error_on_flush()) {
+            return ss::make_exception_future("flush error simulation");
+        }
+        return ss::make_ready_future();
+    }
+
+    void reconcile_properties() final {}
+
+    ss::future<std::optional<translated_offset_range>>
+    finish(retry_chain_node&, ss::abort_source&) final {
+        _inflight_translation = false;
+        translated_offset_range result;
+        if (_min_offset_translated && _max_offset_translated) {
+            result.start_offset = _min_offset_translated.value();
+            result.last_offset = _max_offset_translated.value();
+        }
+        return ss::make_ready_future<std::optional<translated_offset_range>>(
+          std::move(result));
+    }
+
+    ss::future<> discard() final {
+        _inflight_translation = false;
+        return ss::make_ready_future();
+    }
+
+private:
+    std::optional<kafka::offset> _min_offset_translated;
+    std::optional<kafka::offset> _max_offset_translated;
+    fake_test_ctx& _test_ctx;
+    bool _inflight_translation{false};
+};
+
+class partition_translator_fixture : public scheduling::scheduler_fixture {
+public:
+    ss::future<> SetUpAsync() override {
+        co_await scheduling::scheduler_fixture::SetUpAsync();
+    }
+
+    std::unique_ptr<scheduling::scheduling_policy>
+    make_scheduling_policy() override {
+        return scheduling::scheduling_policy::make_default(
+          max_concurrent_translators, task_quota);
+    }
+
+protected:
+    ss::future<bool> add_translator(fake_test_ctx& ctx) {
+        return _scheduler->add_translator(
+          std::make_unique<partition_translator>(
+            test_sg,
+            std::make_unique<fake_coordinator_api>(ctx),
+            std::make_unique<fake_data_src>(ctx),
+            std::make_unique<fake_translation_ctx>(ctx),
+            retry_max_timeout,
+            retry_initial_backoff));
+    }
+
+private:
+    static constexpr auto task_quota
+      = std::chrono::duration_cast<scheduling::clock::duration>(2ms);
+};
+
+TEST_F_CORO(partition_translator_fixture, test_cleanup_on_error) {
+    auto test_ctx = make_test_context();
+    if (tests::random_bool()) {
+        test_ctx.set_error_on_translation(true);
+    } else {
+        test_ctx.set_error_on_flush(true);
+    }
```

Comment:
nit: kinda seems like we would be better off with separate test cases vs flipping between the two at random

Reply:
sure, let me split into a separate test.